### PR TITLE
Update TROUBLESHOOTING.md

### PR DESCRIPTION
**Pull request by msaroufim**
I cut a significant amount of the doc which I feel bad about but if I'm being honest I would space out while reading the existing TROUBLESHOOTING.md and couldn't logically group the different debugging tools. Hopefully this helps

### CODE REVIEW

**Review by yanboliang — COMMENTED** [review]
thanks for the compressing. I think the new version is more actionable but may lose some context or background. It likes a checklist before asking help but users may don't know how to do for each of the step if they are new hand. 
I'm wondering if we can keep the original version at somewhere else for reference and knowledge sharing.

**Comment by msaroufim** [comment]
> thanks for the compressing. I think the new version is more actionable but may lose some context or background. It likes a checklist before asking help but users may don't know how to do for each of the step if they are new hand. I'm wondering if we can keep the original version at somewhere else for reference and knowledge sharing.

Yeah I feel like the existing troubleshooting.md was something in between an internals.md and troubleshooting.md, I did want to take a stab at writing an internals.md but after making troubleshooting more actionable

File: documentation/TROUBLESHOOTING.md [hunk 1]
```diff
@@ -1,347 +1,81 @@
...
-- Setting the environment variable TORCHDYNAMO_DEBUG_FUNCTION to the desired function name will only run torchdynamo on functions with that name.
-- There is a record/replay tool (set `torch._dynamo.config.replay_record_enabled = True`) which dumps an execution record when an error is encountered. This record can then be replayed to run only the frame where an error occurred.
+If a model is sufficiently large, the logs can become overwhelming. If an error occurs deep within a model's python code, it can be useful to execute only the frame in which the error occurs to enable easier debugging. There are 2 tools available to enable this:
+1. `env TORCHDYNAMO_DEBUG_FUNCTION=<desired_function_name>` will only run torchdynamo on functions with that name.
+2. `env torch._dynamo.config.replay_record_enabled = True`) which dumps an execution record when an error is encountered. This record can then be replayed to run only the frame where an error occurred.
 
 
 ## TorchInductor Errors
```
[the remark this answers is not in the document]
Comment (msaroufim) [reply documentation/TROUBLESHOOTING.md:30]: yeah wasn't my intent to remove that, I'll add it back

**Comment by yanboliang** [comment]
> Yeah I feel like the existing troubleshooting.md was something in between an internals.md and troubleshooting.md. I can keep it as is but rename it to internals.md and then call the new doc troubleshooting.md

Agree, one user guide and another developer or advanced user guide.

**Comment by mlazos** [comment]
I think you guys have reached the conclusion I would propose already, I'm fine with multiple documents/breaking this up if necessary if it is too dense.

I think the north star should be these documents:
- Readme/Introduction
- Quick Start + detailed tutorial which touches on the main concepts (graph breaks, JIT, levels of the stack)
- FAQ (I think common error questions like "why is my code slow" should be here)
- Debug Guide (if FAQ doesn't address beginner questions, separating this into power user/beginner is fine)

The debug guide should include
1. Debug walkthroughs for a bug in each level of the stack/each type of bug
2. Detailed example of each tool
3. A table of quick shortcuts to show the usage of each tool for people who don't want to read the document
4. A ToC since it will probably be longer

In short I think this PR deletes way too much useful information, here's the [original context](https://github.com/pytorch/torchdynamo/pull/1173) for the first draft to see why certain things were included.

**Comment by msaroufim** [comment]
Thanks for the feedback @mlazos, would you be more amenable to this change if I instead kept the existing `troubleshooting.md` as is and created a new doc called `faq.md`?

Here is my thinking of what the user journey looks like

1. `README.md`: The vast majority of our users will want to try out dynamo/inductor and see whether they get speedups. For those users their first experience will be going through the README and a 1 line quickstart `torchdynamo.optimize()` and if their model is fast great and if not they go to step 2 
2. `FAQ.md`: Users are debugging a failure and any setback will be one reason why they will stop using dynamo/inductor so focusing on the specific kinds of issues they are seeing and proposing a checklist of sort will keep them motivated to unblock themselves. For these users learning about the dynamo/inductor internals is not a priority and they'd prefer a short `faq.md` that they can quickly skim
3. `INTERNALS.md`: Finally as users are more invested in the ecosystem and maybe they'd like to contribute some other tools will help in better understanding how dynamo/inductor work those tools are ideally suited to a `tooling.md` or `internals.md` or `troubleshooting.md` that help people understand the various layers of the stack by visualizing problems in a specific component. This can be long because people reading this document are opting into learning about nuances in our stack.

**Comment by mlazos** [comment]
Hmm there's a pretty substantial overlap between the documents. I think my main concern with the content getting deleted before is that it is super useful IMO to have full examples, maybe keep those separately and link to them in the new FAQ?

For the FAQ I was thinking of having a list of common questions. For example, what are graph breaks? what are guards? at a high level how does jit work? Nested dynamo calls? Perhaps remove the redundant sections from troubleshooting.md, and make the troubleshooting.md into tooling.md or something. If it's easier to just leave the existing documents, and you're planning future PRs in this area that's fine too.

**Comment by msaroufim** [comment]
Yeah the second paragraph resonates with me

**Review by soumith — APPROVED** [review]
Approving to commit first. And will make small improvements in subsequent PR.
